flux-jobs: add support for job expiration and remaining time

grondo opened the pull request:
This PR adds support for collecting job `expiration` time from _R_ in job-info, along with the changes necessary to support the new fields `expiration` and `t_remaining` in `flux-jobs`.

So that it wasn't necessary to add `t_remaining_fsd` and `t_remaining_hms` etc fields in flux-jobs, support for custom "conversion specifiers" was added to the `JobFormatter` class. Now, a `!F` converts a float time to FSD, `!H` converts to H:M:S, and for datetime objects `!D` converts to an ISO 8601 date and time string (no timezone).
A `!d` was also added for flexibility to convert to a `datetime` object, which then has its own custom format spec (`strftime()` based):
```
$ flux jobs -ao '{t_inactive!d:%Y-%m-%d}'
T_INACTIVE
2020-06-15
2020-06-15
2020-06-15
```
However, this form doesn't allow for alignment so the `!D` is probably more useful in most circumstances.

As a result `runtime_fsd` and `runtime_hms` fields are removed in preference for `runtime!F` and `runtime!H`.

```
$ flux jobs -ao '{id:>20} {t_remaining!F:>12h}'
               JOBID  T_REMAINING
       2378002595840       18.77s
        318347673600       23.97h
        129754988544       2.746m
       1836585058304            -
        176026550272            -
```

Leaving this as a WIP to get comments on the general approach, naming of fields, etc.

garlick commented:
Poking at this and it seems to work just fine.

Don't forget to update flux-jobs(1) `OUTPUT FORMAT` section.

grondo commented:
Ok, forced a push with updated `flux-jobs(1)` manpage (thanks for the reminder @garlick) and slightly expanded tests.

This is probably ready for consideration, so I've removed the WIP label.

grondo commented:
Tried to force push a fix for:

```
src/cmd/flux-jobs.py:177:11: W0143: Comparing against a callable, did you omit the parenthesis? (comparison-with-callable)
```

That line was:
```python
        status = self.status
        if status != "RUNNING":
            return 0.0
```

`self.status` is declared as a property, so I'm not sure why `pylint` is marking that as bug. I tried using `str()` -- we'll see if that works.

grondo commented:
I keep getting hangs in the Ubuntu 20.04 builder on this PR. However, AFAICT, the hanging test is `python/t0009-security.py`, which is clearly unrelated (I hope). I restarted and we'll see if it fails again.

grondo commented:
This PR seems to be causing hangs in Travis. I'll keep looking.

grondo commented:
Fixed one issue in a test, but it wasn't clear why that would cause a hang.

Rebased after #3009 was merged and things appear to be more stable now.
(I wonder if running only 1 job-info module per instance is helping the efficiency of the whole testsuite)

chu11 commented on t/t2230-job-info-list.t:
```diff
@@ -748,6 +748,17 @@ test_expect_success HAVE_JQ 'flux job list outputs exceptions correctly (excepti
+	jq -e ".expiration > now" < expiration.json &&
+	flux job cancel $jobid
+'
+
```
wasn't sure where to leave this comment, but down below in this file there's also a test that the `list-attrs` command is listing all attributes correctly.

chu11 commented on src/cmd/flux-jobs.py:
```diff
@@ -627,7 +645,7 @@ def main():
     else:
         fmt = (
-            "{id:>18} {username:<8.8} {name:<10.10} {status_abbrev:>6.6} "
+            "{id:>20} {username:<8.8} {name:<10.10} {status_abbrev:>6.6} "
```
the increase in id width probably in wrong commit? (flux-jobs: add output fields related to job expiration )

grondo replied:
Hm, I don't even remember making that change... I will nix it.

grondo commented:
Ok, I pushed a couple commits that add a couple extra tests for `!d` and `!D` conversion specifiers, and also rework the `list-attrs` checks to verify against a global list of expected attributes.


I don't want to wild testing every iteration of `!d` formatting at this point. Before that, I'd like to move the `JobInfo` class into the public bindings, at which point unit tests for the class could be added (somewhere).

grondo commented:
If things look ok, I'll squash the fixup commits.

grondo commented:
Ok, I squashed and forced a push.

I also removed that strange change in the default `{id}` output width. Thanks for catching that @chu11!

grondo commented:
Thanks!
